server/statistics/region_collection.go

```diff
@@ -169,17 +191,24 @@ func (r *RegionStatistics) Observe(region *core.RegionInfo, stores []*core.Store
 		}
 	}
 
+	// Better to make sure once any of these conditions changes, it will trigger the heartbeat `save_cache`.
+	// Otherwise, the state may be out-of-date for a long time, which needs another way to apply the change ASAP.
+	// For example, see `RegionStatsNeedUpdate` above to know how `OversizedRegion` and ``UndersizedRegion` are updated.
 	conditions := map[RegionStatisticType]bool{
 		MissPeer:    len(region.GetPeers()) < desiredReplicas,
 		ExtraPeer:   len(region.GetPeers()) > desiredReplicas,
 		DownPeer:    len(region.GetDownPeers()) > 0,
 		PendingPeer: len(region.GetPendingPeers()) > 0,
 		LearnerPeer: len(region.GetLearners()) > 0,
 		EmptyRegion: region.GetApproximateSize() <= core.EmptyRegionApproximateSize,
-		OversizedRegion: region.GetApproximateSize() >= int64(r.storeConfigManager.GetStoreConfig().GetRegionMaxSize()) ||
-			region.GetApproximateKeys() >= int64(r.storeConfigManager.GetStoreConfig().GetRegionMaxKeys()),
-		UndersizedRegion: region.NeedMerge(int64(r.opt.GetScheduleConfig().MaxMergeRegionSize),
-			int64(r.opt.GetScheduleConfig().MaxMergeRegionKeys)),
+		OversizedRegion: region.IsOversized(
+			int64(r.storeConfigManager.GetStoreConfig().GetRegionMaxSize()),
+			int64(r.storeConfigManager.GetStoreConfig().GetRegionMaxKeys()),
+		),
+		UndersizedRegion: region.IsUndersized(
```

Comment:
When the TTL config exists, I'm not sure whether the original config or TTL config should be used.

Reply:
Since the TTL config will affect the actual merge behavior, I think the statistical info should be consistent with it.